File: Java/src/test/java/com/xurxodev/stringcalculator/StringCalculatorShould.java
package com.xurxodev.stringcalculator;

import com.xurxodev.stringcalculator.rule.NegativeNumberValidateRule;
import com.xurxodev.stringcalculator.rule.NumberGreaterThanIgnoredRule;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;

public class StringCalculatorShould {
    @Test
    public void return_zero_if_input_is_empty() throws Exception {
        int expectedResult = 0;

        StringCalculator calculator = givenAStringCalculator();

        int result = calculator.add("");

        assertThat(result, is(expectedResult));
    }

    @Test
    public void return_one_if_input_is_one() throws Exception {
        int expectedResult = 1;

        StringCalculator calculator = givenAStringCalculator();

        int result = calculator.add("1");

        assertThat(result, is(expectedResult));
    }

    @Test
    public void return_three_if_input_is_one_and_two() throws Exception {
        int expectedResult = 3;

        StringCalculator calculator = givenAStringCalculator();

        int result = calculator.add("1,2");

        assertThat(result, is(expectedResult));
    }

    @Test
    public void return_fifteen_if_input_is_from_one_to_five() throws Exception {
        int expectedResult = 15;

        StringCalculator calculator = givenAStringCalculator();

        int result = calculator.add("1,2,3,4,5");

        assertThat(result, is(expectedResult));
    }

    @Test
    public void return_two_if_input_is_two_and_thousand_and_one() throws Exception {
        int expectedResult = 2;

        StringCalculator calculator = givenAStringCalculator();

        int result = calculator.add("2,1001");

        assertThat(result, is(expectedResult));
    }

    @Rule
    public ExpectedException thrown = ExpectedException.none();

    @Test
    public void throw_exception_if_input_contains_negative_number() throws Exception {
        thrown.expect(Exception.class);
        thrown.expectMessage("Negatives not allowed");

        StringCalculator calculator = givenAStringCalculator();

        int result = calculator.add("2,-5");
    }

    private StringCalculator givenAStringCalculator() {
        StringCalculator.NumberExtractor numberExtractor = new NumberExtractorFromCommaSeparatedString();
        StringCalculator.NumberRemover numberRemover = createNumberRemover();
        StringCalculator.NumberValidator numberValidator = createNumberValidator();

        return new StringCalculator(numberExtractor,numberRemover,numberValidator);
    }

    private StringCalculator.NumberRemover createNumberRemover(){
        final NumberRemoverByIgnoredRules.IgnoredRule ignoredRule = new NumberGreaterThanIgnoredRule(1000);

        List<NumberRemoverByIgnoredRules.IgnoredRule> ignoredRules = new ArrayList<>();

        ignoredRules.add(ignoredRule);

        return new NumberRemoverByIgnoredRules(ignoredRules);
    }

    private StringCalculator.NumberValidator createNumberValidator(){
        final NumberValidatorByValidateRules.ValidateRule validateRule = new NegativeNumberValidateRule();

        List<NumberValidatorByValidateRules.ValidateRule> validateRules = new ArrayList<>();

        validateRules.add(validateRule);

        return new NumberValidatorByValidateRules(validateRules);
    }
}

File: C#/StringCalculatorKata.Tests/StringCalculatorShould.cs
using StringCalculatorKata.Boundaries;
using StringCalculatorKata.Errors;
using StringCalculatorKata.Rules;
using System;
using System.Collections.Generic;
using Xunit;

namespace StringCalculatorKata.Tests
{
    public class StringCalculatorShould
    {
        [Fact]
        public void return_zero_if_input_is_empty()
        {
            int expectedResult = 0;

            StringCalculator calculator = GivenAStringCalculator(); 

            int result = calculator.Add("");

            Assert.Equal(expectedResult, result);
        }

        [Fact]
        public void return_one_if_input_is_one()
        {
            int expectedResult = 1;

            StringCalculator calculator = GivenAStringCalculator();

            int result = calculator.Add("1");

            Assert.Equal(expectedResult, result);
        }

        [Fact]
        public void return_three_if_input_is_one_and_two()
        {
            int expectedResult = 3;

            StringCalculator calculator = GivenAStringCalculator();

            int result = calculator.Add("1,2");

            Assert.Equal(expectedResult, result);
        }

        [Fact]
        public void return_fifteen_if_input_is_from_one_to_five()
        {
            int expectedResult = 15;

            StringCalculator calculator = GivenAStringCalculator();

            int result = calculator.Add("1,2,3,4,5");

            Assert.Equal(expectedResult, result);
        }

        [Fact]
        public void return_two_if_input_is_two_and_thousand_and_one()
        {
            int expectedResult = 2;

            StringCalculator calculator = GivenAStringCalculator();

            int result = calculator.Add("2,1001");

            Assert.Equal(expectedResult, result);
        }

        [Fact]
        public void throw_exception_if_input_contains_negative_number()
        {
            NegativeNumbersNotAllowedException ex = Assert.Throws<NegativeNumbersNotAllowedException>(() =>
            {
                StringCalculator calculator = GivenAStringCalculator();

                int result = calculator.Add("2,-5");
            });

            Assert.Equal("Negatives not allowed", ex.Message);
        }

        private StringCalculator GivenAStringCalculator()
        {
            INumberExtractor numberExtractor = new NumberExtractorFromCommaSeparatedString();
            INumberRemover numberRemover = createNumberRemover();
            INumberValidator numberValidator = createNumberValidator();

            return new StringCalculator(numberExtractor, numberRemover, numberValidator);
        }

        private INumberRemover createNumberRemover()
        {
            IIgnoreRule ignoredRule = new NumberGreaterThanIgnoredRule(1000);

            List<IIgnoreRule> ignoredRules = new List<IIgnoreRule>();

            ignoredRules.Add(ignoredRule);

            return new NumberRemoverByIgnoredRules(ignoredRules);
        }

        private INumberValidator createNumberValidator()
        {
            IValidateRule validateRule = new NegativeNumberValidateRule();

            List<IValidateRule> validateRules = new List<IValidateRule>();

            validateRules.Add(validateRule);

            return new NumberValidatorByValidateRules(validateRules);
        }
    }
}
